cmd/codegen/generator/go/templates/module_interfaces.go

```diff
@@ -0,0 +1,540 @@
+		}
+
+		if !goFuncType.Exported() {
+			continue
+		}
+
+		goFuncTypes = append(goFuncTypes, goFuncType)
+	}
+	// verify the DaggerObject interface methods are all there
+	for methodObj, found := range daggerObjectIfaceMethods {
+		if !found {
+			return nil, fmt.Errorf("missing method %s from DaggerObject interface, which must be embedded in interfaces used in Functions and Objects", methodObj.Name())
+		}
+	}
+
+	sort.Slice(goFuncTypes, func(i, j int) bool {
+		return goFuncTypes[i].Pos() < goFuncTypes[j].Pos()
+	})
+	for _, goFuncType := range goFuncTypes {
+		funcTypeSpec, err := ps.parseGoFunc(named, goFuncType)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse method %s: %w", goFuncType.Name(), err)
+		}
+		spec.methods = append(spec.methods, funcTypeSpec)
+	}
+
+	// get the comment above the interface (if any)
+	astSpec, err := ps.astSpecForNamedType(named)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find decl for named type %s: %w", spec.name, err)
+	}
+	spec.doc = astSpec.Doc.Text()
+
+	return spec, nil
+}
+
+type parsedIfaceType struct {
+	name string
+	doc  string
+
+	methods []*funcTypeSpec
+
+	goType     *types.Interface
+	moduleName string
+}
+
+var _ NamedParsedType = &parsedIfaceType{}
+
+func (spec *parsedIfaceType) TypeDefCode() (*Statement, error) {
+	withIfaceArgsCode := []Code{
+		Lit(spec.name),
+	}
+	withIfaceOptsCode := []Code{}
+	if spec.doc != "" {
+		withIfaceOptsCode = append(withIfaceOptsCode, Id("Description").Op(":").Lit(strings.TrimSpace(spec.doc)))
+	}
+	if len(withIfaceOptsCode) > 0 {
+		withIfaceArgsCode = append(withIfaceArgsCode, Id("TypeDefWithInterfaceOpts").Values(withIfaceOptsCode...))
+	}
+
+	typeDefCode := Qual("dag", "TypeDef").Call().Dot("WithInterface").Call(withIfaceArgsCode...)
+
+	for _, method := range spec.methods {
+		fnTypeDefCode, err := method.TypeDefCode()
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert method %s to function def: %w", method.name, err)
+		}
+		typeDefCode = dotLine(typeDefCode, "WithFunction").Call(Add(Line(), fnTypeDefCode))
+	}
+
+	return typeDefCode, nil
+}
+
+func (spec *parsedIfaceType) GoType() types.Type {
+	return spec.goType
+}
+
+func (spec *parsedIfaceType) GoSubTypes() []types.Type {
+	var subTypes []types.Type
+	for _, method := range spec.methods {
+		subTypes = append(subTypes, method.GoSubTypes()...)
+	}
+	return subTypes
+}
+
+func (spec *parsedIfaceType) Name() string {
+	return spec.name
+}
+
+func (spec *parsedIfaceType) ImplementationCode() (*Statement, error) {
+	code := Empty().
+		Add(spec.concreteStructDefCode()).Line().
+		Add(spec.idDefCode()).Line().
+		Add(spec.loadFromIDMethodCode()).Line().
+		Add(spec.graphqlTypeMethodCode()).Line().
+		Add(spec.graphqlIDTypeMethodCode()).Line().
+		Add(spec.graphqlIDMethodCode()).Line().
+		Add(spec.marshalJSONMethodCode()).Line().
+		Add(spec.unmarshalJSONMethodCode()).Line().
+		Add(spec.toIfaceMethodCode()).Line()
```

Comment:
Ahhh, I hate being picky without specific suggestions, but...

This structure of code is not particularly easy to understand, especially some of the deeply chained jennifer call later on in this file (notably `concreteMethodExecuteQueryCode`). Given this code has been proven to be quite delicate in the past, I'm a bit worried of subtle bugs sneaking in at some point. The thought of debugging differences between templates and jennifer having different parts of code does worry me a bit if/when things crop up.

On the other hand, everything else in this PR is really not hard to understand, and the test suite does give me confidence that we've not broken *everything*. I *really* don't want to block this PR, but I think we should try and tidy this up sooner rather than later (whether that's bringing the templated code into jennifer, or jennifer into the templated code, etc).

Reply:
Yeah totally, I did the bare minimum cleanup of this code (it was even worse in the first iteration) and then briefly embarked on even more cleanup but just had to back out due to time. I think I see various paths to making it more readable by basically adding *a lot* more jennifer utils, but I quickly realized that was gonna add another day or two to implementation time, which given our current deadlines was too much.

That's also why I didn't embark on trying to unify everything to jennifer across the whole go sdk (which would be my preference relative to unifying everything to use templates). I will create a follow-up issue, but it's going to be hard to prioritize in the next month I think 😞.

> The thought of debugging differences between templates and jennifer having different parts of code does worry me a bit if/when things crop up.

Yeah I have the same concern, but my suspicion was that the alternative of trying to centralize everything to use templates would have:
1. Been approximately as confusing + convoluted (the template related code is obviously very hard to work with) and would require subtle adjustments to the templates used for go sdk client codegen (which increases blast radius)
2. Not given us this setup for replacing all the go template code in the client-side w/ jennifer, which I think is the best long term option

So my hope is that this approach at least serves as a bridge to a better future even if there's some semblance of duplication in the meantime

The other consolation here besides the tests I added is that this all only triggers when a Go SDK module uses an interface in its typedefs, so at least its blast radius is also more scoped in that respect too.